### PR TITLE
- Added support for DataFrame in ScalaSparkCompute

### PR DESCRIPTION
**Pull request by chtyim**
- Added support for extra dependencies in ScalaSparkProgram

### CODE REVIEW

File: docs/ScalaSparkCompute-sparkcompute.md [hunk 1]
```diff
@@ -14,17 +14,23 @@ of the join result using Spark SQL.
 Properties
 ----------
 **scalaCode** Spark code in Scala defining how to transform RDD to RDD. 
-The code must implement a function called ``transform``, which the signature should either be
+The code must implement a function called ``transform``, which the signature should one of the follow form:
```
Comment (albertshau) [inline docs/ScalaSparkCompute-sparkcompute.md]: 'which the signature should one of the follow form:' -> 'whose signature should be one of:'

File: docs/ScalaSparkCompute-sparkcompute.md [hunk 2]
```diff
@@ -14,17 +14,23 @@ of the join result using Spark SQL.
...
-    def transform(rdd: RDD[StructuredRecord], context: SparkExecutionPluginContext) : RDD[StructuredRecord]
+    def transform(df: DataFrame, context: SparkExecutionPluginContext) : DataFrame
     
-In the latter case, using the ``SparkExecutionPluginContext`` can access CDAP
+The input ``DataFrame`` has the same schema as the input schema to this stage and the ``transform`` method
+should return a ``DataFrame`` that has the same schema as the output schema setup for this stage.
+Using the ``SparkExecutionPluginContext`` can access CDAP
 entities such as Stream and Dataset, as well as providing access to the underlying ``SparkContext`` in use.
+ 
+Operating on lower level ``RDD`` is also possible by using the one of the following form of the ``transform`` method:
```
Comment (albertshau) [inline docs/ScalaSparkCompute-sparkcompute.md]: form -> forms

File: docs/ScalaSparkCompute-sparkcompute.md [hunk 3]
```diff
@@ -14,17 +14,23 @@ of the join result using Spark SQL.
...
-or
+    def transform(df: DataFrame) : DataFrame
 
-    def transform(rdd: RDD[StructuredRecord], context: SparkExecutionPluginContext) : RDD[StructuredRecord]
+    def transform(df: DataFrame, context: SparkExecutionPluginContext) : DataFrame
     
-In the latter case, using the ``SparkExecutionPluginContext`` can access CDAP
+The input ``DataFrame`` has the same schema as the input schema to this stage and the ``transform`` method
+should return a ``DataFrame`` that has the same schema as the output schema setup for this stage.
+Using the ``SparkExecutionPluginContext`` can access CDAP
```
Comment (albertshau) [inline docs/ScalaSparkCompute-sparkcompute.md]: Using the SparkExecutionPluginContext, you can access

File: src/main/java/co/cask/hydrator/plugin/spark/dynamic/SparkDataFrames.java [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        List<StructField> structFields = new ArrayList<>();
+        for (Schema.Field field : schema.getFields()) {
+          Schema fieldSchema = field.getSchema();
+          DataType fieldType = toDataType(fieldSchema.isNullable() ? fieldSchema.getNonNullable() : fieldSchema);
+          structFields.add(DataTypes.createStructField(field.getName(), fieldType, fieldSchema.isNullable()));
+        }
+        return DataTypes.createStructType(structFields);
+      case UNION:
+        if (schema.isNullable()) {
+          return toDataType(schema.getNonNullable());
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/plugin/spark/dynamic/SparkDataFrames.java:133]: there seems to be a mismatch between converting cdap schema to spark datatype and back, where a nullable string in cdap would become a string datatype in spark, which would become a non-nullable string in cdap.

File: src/main/java/co/cask/hydrator/plugin/spark/dynamic/SparkDataFrames.java [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        return DataTypes.BinaryType;
+      case STRING:
+        return DataTypes.StringType;
+      case ENUM:
+        return DataTypes.StringType;
+      case ARRAY:
+        return DataTypes.createArrayType(toDataType(schema.getComponentSchema()));
+      case MAP:
+        Map.Entry<Schema, Schema> mapSchema = schema.getMapSchema();
+        return DataTypes.createMapType(toDataType(mapSchema.getKey()), toDataType(mapSchema.getValue()));
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/plugin/spark/dynamic/SparkDataFrames.java]: seems like we should be using the 3 argument createMapType(), with the valueContainsNull argument, unless that's only in the spark 2.1 version im looking at.

File: src/main/java/co/cask/hydrator/plugin/spark/dynamic/SparkDataFrames.java [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+      case DOUBLE:
+        return DataTypes.DoubleType;
+      case BYTES:
+        return DataTypes.BinaryType;
+      case STRING:
+        return DataTypes.StringType;
+      case ENUM:
+        return DataTypes.StringType;
+      case ARRAY:
+        return DataTypes.createArrayType(toDataType(schema.getComponentSchema()));
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/plugin/spark/dynamic/SparkDataFrames.java]: there's also a createArrayType() with a 'containsNull' boolean argument, unless that's only in the spark 2.1 version im looking at.

File: widgets/ScalaSparkProgram-sparkprogram.json [hunk 7]
```diff
@@ -21,6 +21,14 @@
           "widget-attributes": {
             "default": "import co.cask.cdap.api.spark._\nimport org.apache.spark._\nimport org.slf4j._\n\nclass SparkProgram extends SparkMain {\n  import SparkProgram._\n\n  override def run(implicit sec: SparkExecutionContext): Unit = {\n    LOG.info(\"Spark Program Started\")\n\n    val sc = new SparkContext\n\n    LOG.info(\"Spark Program Completed\")\n  }\n}\n\nobject SparkProgram {\n  val LOG = LoggerFactory.getLogger(getClass())\n}"
           }
+        },
+        {
+          "widget-type": "dsv",
```
Comment (albertshau) [inline widgets/ScalaSparkProgram-sparkprogram.json:26]: csv?

Comment (chtyim) [reply widgets/ScalaSparkProgram-sparkprogram.json:26]: It's dsv (delimiter separated).

File: src/main/java/co/cask/hydrator/plugin/spark/dynamic/SparkDataFrames.java (hunk 4 above)
Comment (chtyim) [reply src/main/java/co/cask/hydrator/plugin/spark/dynamic/SparkDataFrames.java:133]: Updated the handling of nullable for array, map and record fields.

File: docs/ScalaSparkCompute-sparkcompute.md (hunk 3 above)
Comment (chtyim) [reply docs/ScalaSparkCompute-sparkcompute.md]: fix

File: docs/ScalaSparkCompute-sparkcompute.md (hunk 1 above)
Comment (chtyim) [reply docs/ScalaSparkCompute-sparkcompute.md]: fixed

File: docs/ScalaSparkCompute-sparkcompute.md (hunk 2 above)
Comment (chtyim) [reply docs/ScalaSparkCompute-sparkcompute.md]: fixed

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Squashed and merging.

**Pull request merged**